src/connector/src/sink/file_sink/opendal_sink.rs

```diff
@@ -0,0 +1,315 @@
+    /// When a checkpoint arrives, the force commit is performed to write the data to the file.
+    /// In the future if flush and checkpoint is decoupled, we should enable sink decouple accordingly.
+    async fn barrier(&mut self, is_checkpoint: bool) -> Result<()> {
+        if is_checkpoint && let Some(sink_writer) = self.sink_writer.take() {
+            match sink_writer {
+                FileWriterEnum::ParquetFileWriter(w) => {
+                    let _ = w.close().await?;
+                }
+            };
+        }
+
+        Ok(())
+    }
+}
+
+impl OpenDalSinkWriter {
+    pub fn new(
+        operator: Operator,
+        write_path: &str,
+        rw_schema: Schema,
+        is_append_only: bool,
+        executor_id: u64,
+        encode_type: SinkEncode,
+    ) -> Result<Self> {
+        let arrow_schema = convert_rw_schema_to_arrow_schema(rw_schema)?;
+        Ok(Self {
+            schema: Arc::new(arrow_schema),
+            write_path: write_path.to_string(),
+            operator,
+            sink_writer: None,
+            is_append_only,
+            epoch: None,
+            executor_id,
+            encode_type,
+        })
+    }
+
+    async fn create_object_writer(&mut self, epoch: u64) -> Result<OpendalWriter> {
+        // Todo: specify more file suffixes based on encode_type.
+        let suffix = match self.encode_type {
+            SinkEncode::Parquet => "parquet",
+            _ => unimplemented!(),
+        };
+
+        // Note: sink decoupling is not currently supported, which means that output files will not be batched across checkpoints.
+        // The current implementation writes files every time a checkpoint arrives, so the naming convention is `epoch + executor_id + .suffix`.
+        let object_name = format!(
+            "{}/{}_{}.{}",
+            self.write_path, epoch, self.executor_id, suffix,
+        );
+        Ok(self
+            .operator
+            .writer_with(&object_name)
+            .concurrent(8)
+            .await?)
+    }
+
+    async fn create_sink_writer(&mut self, epoch: u64) -> Result<()> {
+        let object_writer = self.create_object_writer(epoch).await?;
+        match self.encode_type {
+            SinkEncode::Parquet => {
+                let props = WriterProperties::builder();
+                let parquet_writer: tokio_util::compat::Compat<opendal::FuturesAsyncWriter> =
+                    object_writer.into_futures_async_write().compat_write();
+                self.sink_writer = Some(FileWriterEnum::ParquetFileWriter(
+                    AsyncArrowWriter::try_new(
+                        parquet_writer,
+                        self.schema.clone(),
+                        Some(props.build()),
+                    )?,
+                ));
+            }
+            // SinkEncode::Json => {
+            //     self.sink_writer = Some(FileWriterEnum::FileWriter(object_writer));
+            //     unimplemented!();
+            // }
```

Comment:
Given that we don't have relevant codes in the code bases, I suggest we remove this line instead of commenting it out.